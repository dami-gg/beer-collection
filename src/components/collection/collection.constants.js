// @flow

export const RESULTS_PER_PAGE: number = 18;
