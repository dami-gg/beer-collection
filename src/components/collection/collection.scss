.collection {
  padding: 15px;

  .beer {
    border: 1px solid dimgrey;
    border-radius: 20px;
    margin: 10px 0;
    text-align: center;
    cursor: pointer;

    @media(min-width: 768px) {
      margin: 10px;
    }

    &:hover {
      background-color: lightskyblue;
    }

    &__image {
      max-height: 100%;
      max-width: 100%;
    }
  }
}
