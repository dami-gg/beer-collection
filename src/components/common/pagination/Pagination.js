// @flow
import React, { PureComponent } from "react";

import PageButton from "./PageButton";

import { getButtonLabels } from "./pagination.helpers";
import { PLACEHOLDER_PAGE_LABEL } from "./pagination.constants";

import "./pagination.scss";

type Props = {
  className: string,
  numItems: number,
  currentPage: number,
  resultsPerPage: number,
  totalPageButtons: number,
  onNavigation: Function
};

class Pagination extends PureComponent<Props> {
  totalPages: number;
  navigateToPage: Function;
  navigateToPreviousPage: Function;
  navigateToNextPage: Function;

  constructor(props: Props) {
    super(props);

    this.navigateToPage = this.navigateToPage.bind(this);
    this.navigateToPreviousPage = this.navigateToPreviousPage.bind(this);
    this.navigateToNextPage = this.navigateToNextPage.bind(this);

    this.totalPages = this.getTotalPages();
  }

  componentWillUpdate(nextProps: Object) {
    this.totalPages = this.getTotalPages(nextProps.numItems);
  }

  getTotalPages(numItems: number = this.props.numItems) {
    return Math.ceil(numItems / this.props.resultsPerPage);
  }

  navigateToPage(page: number) {
    if (page !== PLACEHOLDER_PAGE_LABEL) {
      this.props.onNavigation(page);
    }
  }

  navigateToPreviousPage() {
    if (this.props.currentPage !== 1) {
      this.navigateToPage(this.props.currentPage - 1);
    }
  }

  navigateToNextPage() {
    if (this.props.currentPage !== this.totalPages) {
      this.navigateToPage(this.props.currentPage + 1);
    }
  }

  generatePageButtons() {
    return getButtonLabels(
      this.totalPages,
      this.props.totalPageButtons,
      this.props.currentPage
    ).map((item, i) => (
      <PageButton
        key={i}
        label={item}
        className={
          Math.abs(i - (this.props.currentPage - 1)) > 1
            ? "small-screens--hidden"
            : ""
        }
        clickHandler={this.navigateToPage}
        active={this.props.currentPage === item}
        disabled={item === PLACEHOLDER_PAGE_LABEL}
      />
    ));
  }

  generateOutmostButton(position: string = "left") {
    if (this.props.numItems && this.props.numItems > 0) {
      const isLeftPositioned = position === "left";
      const isDisabled = isLeftPositioned
        ? this.props.currentPage === 1
        : this.props.currentPage === this.totalPages;

      return (
        <PageButton
          label={
            isLeftPositioned ? (
              <i className={`fa fa-arrow-left`} aria-hidden="true" />
            ) : (
              <i className={`fa fa-arrow-right`} aria-hidden="true" />
            )
          }
          disabled={isDisabled}
          clickHandler={
            isLeftPositioned
              ? this.navigateToPreviousPage
              : this.navigateToNextPage
          }
        />
      );
    }
  }

  render() {
    return (
      <div
        className={`pagination ${this.totalPages === 1
          ? "pagination--hidden"
          : ""} ${this.props.className}`}>
        {this.generateOutmostButton("left")}

        {this.generatePageButtons()}

        {this.generateOutmostButton("right")}
      </div>
    );
  }
}

export default Pagination;
