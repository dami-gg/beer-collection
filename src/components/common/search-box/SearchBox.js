// @flow
import React, { Component } from "react";

import "./search-box.scss";

type State = {
  value: string
};

class SearchBox extends Component {
  state: State;
  onChange: Function;
  clearInput: Function;

  props: {
    className: string,
    changeHandler: Function
  };

  constructor(props: Object) {
    super(props);

    this.state = {
      value: ""
    };

    this.onChange = this.onChange.bind(this);
    this.clearInput = this.clearInput.bind(this);
  }

  onChange(event: Object): void {
    const value = event.target.value;

    this.setState({ value });
    this.props.changeHandler(value);
  }

  inputIsDirty(): boolean {
    return this.state.value !== "";
  }

  clearInput(): void {
    this.setState({ value: "" });
    this.props.changeHandler("");
  }

  render() {
    return (
      <div className={`search-box ${this.props.className}`}>
        <input
          className="search-box__input"
          type="text"
          value={this.state.value}
          placeholder="Search by name"
          onChange={this.onChange}
        />
        <div className="search-box__clear">
          {this.inputIsDirty() &&
            <i
              className="search-box__clear__button fa fa-times"
              onClick={this.clearInput}
            />}
        </div>
        <i className="search-box__icon fa fa-search" />
      </div>
    );
  }
}

export default SearchBox;
