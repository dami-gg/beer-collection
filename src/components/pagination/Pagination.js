// @flow
import type {Beer } from '../../types';

import React, { Component } from 'react';
import { connect } from 'react-redux';

import { RESULTS_PER_PAGE } from '../../constants';

class Pagination extends Component {
    props: {
        collection: Array<Beer>,
        currentPage: number
    };

    getTotalPages() {
        return this.props.collection ? (this.props.collection.length / RESULTS_PER_PAGE) + 1 : 0;
    }

    getPagination(currentPage: number) {
        let totalPages = 3; // TODO this.getTotalPages();

        return [...Array(totalPages)].map((item, i) =>
            `<li className="${currentPage === i ? 'active' : ''}" key="${i}">${i}</li>`
        );
    }

    render() {
        return (
            <ul className="pagination">
                {
                    this.getTotalPages() > 1 &&
                    this.getPagination(this.props.currentPage)
                }
            </ul>
        );
    }
}

const mapStateToProps = (state: Object) => ({
    collection: state.collection,
    currentPage: state.currentPage || 1 // TODO
});

export default connect(mapStateToProps)(Pagination);