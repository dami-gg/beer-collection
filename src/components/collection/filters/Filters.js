// @flow
import React, {PureComponent} from 'react';

import SearchBox from '../../common/search-box/SearchBox';
import Pagination from '../../common/pagination/Pagination';

const TOTAL_PAGE_BUTTONS: number = 7;

import {getRegularExpression} from './filters.helpers';

class Filters extends PureComponent {
  props: {
    numItems: number,
    resultsPerPage: number,
    currentPage: number,
    onFilterUpdate: Function,
    onPageChange: Function
  };

  updateFilter = (event: Object) => {
    const regex = getRegularExpression(event.target.value);
    this.props.onFilterUpdate(regex);
  }

  render() {
    return (
        <div className="collection__filters">
          <SearchBox
              className="collection__filters__search"
              changeHandler={this.updateFilter}>
          </SearchBox>

          <Pagination
              className="collection__filters__pagination"
              numItems={this.props.numItems}
              currentPage={this.props.currentPage}
              resultsPerPage={this.props.resultsPerPage}
              totalPageButtons={TOTAL_PAGE_BUTTONS}
              onNavigation={this.props.onPageChange}>
          </Pagination>
        </div>
    );
  }
}

export default Filters;
