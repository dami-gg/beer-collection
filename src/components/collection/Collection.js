// @flow
import type { Beer } from "../../types/beer.types";

type State = {
  searchFilterRegex?: Object,
  typeFilter?: string,
  originFilter?: string,
  currentPage: number
};

import React, { Component } from "react";
import { connect } from "react-redux";
import { withRouter } from "react-router";

import Filters from "./filters/Filters";
import Results from "./results/Results";
import FloatingButton from "../common/floating-button/FloatingButton";

import {getAllBeerTypes, getAllBeerOrigins} from '../../helpers/collection.helpers';
import "./collection.scss";

const RESULTS_PER_PAGE: number = 18;

export class Collection extends Component {
  state: State;
  navigateToPage: Function;
  updateSearchFilterRegex: Function;
  redirectToAddPage: Function;
  updateTypeFilter: Function;
  allBeerTypes: Array<string>;
  allBeerOrigins: Array<string>;

  props: {
    collection: Array<Beer>,
    match: Object,
    location: Object,
    history: Object
  };

  constructor(props: Object) {
    super(props);

    this.state = {
      searchFilterRegex: undefined,
      typeFilter: undefined,
      originFilter: undefined,
      currentPage: 1
    };

    this.navigateToPage = this.navigateToPage.bind(this);
    this.updateSearchFilterRegex = this.updateSearchFilterRegex.bind(this);
    this.redirectToAddPage = this.redirectToAddPage.bind(this);
    this.updateTypeFilter = this.updateTypeFilter.bind(this);
  }

  componentWillMount(): void {
    this.allBeerTypes = getAllBeerTypes(this.props.collection);
    this.allBeerOrigins = getAllBeerOrigins(this.props.collection);
  }

  updateSearchFilterRegex(searchFilterRegex: Object): void {
    this.setState({ searchFilterRegex });
  }

  navigateToPage = (currentPage: number): void => {
    this.setState({ currentPage });
  };

  redirectToAddPage = (): void => {
    this.props.history.push("/beer/add");
  };

  updateTypeFilter = (type: string): void => {
    this.setState({typeFilter: type});
  };

  updateOriginFilter = (origin: string): void => {
    this.setState({originFilter: origin});
  };

  render() {
    return (
      <div className="collection">
        <Filters
          numItems={this.props.collection.length}
          resultsPerPage={RESULTS_PER_PAGE}
          currentPage={this.state.currentPage}
          onSearchFilterUpdate={this.updateSearchFilterRegex}
          onPageChange={this.navigateToPage}
          allBeerTypes={this.allBeerTypes}
          onTypeFilterUpdate={this.updateTypeFilter}
          allBeerOrigins={this.allBeerOrigins}
          onOriginFilterUpdate={this.updateOriginFilter}
        />

        <Results
          collection={this.props.collection}
          searchFilterRegex={this.state.searchFilterRegex}
          resultsPerPage={RESULTS_PER_PAGE}
          currentPage={this.state.currentPage}
          allBeerTypes={this.allBeerTypes}
          typeFilter={this.state.typeFilter}
          allBeerOrigins={this.allBeerOrigins}
          originFilter={this.state.originFilter}
        />

        <FloatingButton
          iconClass="fa fa-plus"
          label="Add a new beer"
          buttonColor="green"
          iconColor="white"
          clickHandler={this.redirectToAddPage}
        />
      </div>
    );
  }
}

const mapStateToProps = (state: Object) => ({
  collection: state.collection
});

export default withRouter(connect(mapStateToProps)(Collection));
