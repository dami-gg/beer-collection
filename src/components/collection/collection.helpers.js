// @flow
import type { Beer } from "../../types/beer.types";

import { RESULTS_PER_PAGE, ALL_FILTER_OPTION } from "./collection.constants";

export const findBeerInCollectionById = (
  id: string,
  collection: Array<Beer>
): any => collection.find((element: Beer) => element.id === id);

export const getAllBeerTypes = (collection: Array<Beer>) =>
  collection.reduce((accumulator: Array<any>, current: Beer) => {
    if (!accumulator.find(element => element === current.type)) {
      accumulator.push(current.type);
    }
    return accumulator;
  }, []);

export const getAllBeerOrigins = (collection: Array<Beer>) =>
  collection.reduce((accumulator: Array<any>, current: Beer) => {
    if (!accumulator.find(element => element === current.origin)) {
      accumulator.push(current.origin);
    }
    return accumulator;
  }, []);

export const getPaginatedBeers = (
  collection: Array<Beer> = [],
  page: number = 1
): Array<Beer> =>
  collection.slice(RESULTS_PER_PAGE * (page - 1), RESULTS_PER_PAGE * page - 1);

export const preloadBeerImages = (beersToPreload: Array<Beer> = []): void => {
  beersToPreload.forEach((beer: Beer) => preloadBeerImage(beer));
};

export const preloadBeerImage = (beer: Beer): void => {
  if (beer && beer.image) {
    let image = new Image();
    image.src = beer.image;
  }
};

export const preloadCollectionImagesForPage = (
  collection: Array<Beer> = [],
  page: number = 1
) => preloadBeerImages(getPaginatedBeers(collection, page));

export const getFilteredBeers = (
  collection: Array<Beer>,
  filterRegex?: Object,
  typeFilter?: string,
  originFilter?: string
): Array<Beer> => {
  return collection.filter((beer: Beer) =>
    applyFilters(beer, filterRegex, typeFilter, originFilter)
  );
};

const applyFilters = (
  beer: Beer,
  filterRegex?: Object,
  typeFilter?: string,
  originFilter?: string
): boolean => {
  if (filterRegex && !filterRegex.test(beer.name)) {
    return false;
  }
  if (
    typeFilter &&
    beer.type !== typeFilter &&
    typeFilter !== ALL_FILTER_OPTION
  ) {
    return false;
  }
  if (
    originFilter &&
    beer.origin !== originFilter &&
    originFilter !== ALL_FILTER_OPTION
  ) {
    return false;
  }

  return true;
};

export const getBeersForPage = (
  beers: Array<Beer>,
  currentPage: number
): Array<Beer> => {
  const offset = (currentPage ? currentPage - 1 : 0) * RESULTS_PER_PAGE;

  return beers.slice(offset, offset + RESULTS_PER_PAGE);
};

const COLUMN_DELIMITER: string = ",";
const LINE_DELIMITER: string = "\r\n";

const convertCollectionToCsv: string = (collection: Array<Beer>) => {
  let csv: string = "";

  collection.forEach((beer: Beer, index: number) => {
    csv += `${beer.id}${COLUMN_DELIMITER}
            ${beer.name}${COLUMN_DELIMITER}
            ${beer.type}${COLUMN_DELIMITER}
            ${beer.origin}${COLUMN_DELIMITER}
            ${beer.image}${COLUMN_DELIMITER}
            ${beer.rating}${LINE_DELIMITER}`;
  });

  return csv;
};

export const exportAsCsvFile: void = (
  collection: Array<Beer>,
  fileName?: string
) => {
  if (!collection || !collection.length) {
    return;
  }

  let csv = "data:text/csv;charset=utf-8," + convertCollectionToCsv(collection);

  let link = document.createElement("a");
  link.setAttribute("href", encodeURI(csv));
  link.setAttribute("download", fileName || "export.csv");
  link.click();
};
