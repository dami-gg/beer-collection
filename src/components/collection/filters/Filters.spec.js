import React from "react";
import Filters from "./Filters";

import { shallow } from "enzyme";
import renderer from "react-test-renderer";

import SearchBox from "../../common/search-box/SearchBox";
import Pagination from "../../common/pagination/Pagination";

import { RESULTS_PER_PAGE } from "../collection.constants";

describe("Filters", () => {
  let component, element, mock;

  beforeEach(mocks);

  it("should render without crashing", () => {
    shallow(<Filters />);
  });

  it("should match the snapshot", () => {
    component = renderer.create(<Filters />);
    let json = component.toJSON();

    expect(json).toMatchSnapshot();
  });

  describe("render", () => {
    beforeEach(setup);

    it("should render a SearchBox component", () => {
      element = component.find(SearchBox);

      expect(element).toBeDefined();
    });

    it("should render a SearchBox component which updates the filter on change", () => {
      element = component.find(SearchBox);

      expect(element.props().changeHandler).toBeDefined();
    });

    it("should render a Pagination component", () => {
      element = component.find(Pagination);

      expect(element).toBeDefined();
    });

    it("should render a Pagination component specifying the total number of items", () => {
      element = component.find(Pagination);

      expect(element.props().numItems).toBe(mock.props.numItems);
    });

    it("should render a Pagination component specifying the current page", () => {
      element = component.find(Pagination);

      expect(element.props().currentPage).toBe(mock.props.currentPage);
    });

    it("should render a Pagination component specifying the results per page", () => {
      element = component.find(Pagination);

      expect(element.props().resultsPerPage).toBe(RESULTS_PER_PAGE);
    });

    it("should render a Pagination component which does a callback on page change", () => {
      element = component.find(Pagination);

      expect(element.props().onNavigation).toBe(mock.props.onPageChange);
    });
  });

  function setup() {
    render();
  }

  function render() {
    component = shallow(
      <Filters
        numItems={mock.props.numItems}
        currentPage={mock.props.currentPage}
        onFilterUpdate={mock.props.onFilterUpdate}
        onPageChange={mock.props.onPageChange}
        allBeerTypes={mock.props.allBeerTypes}
        allBeerOrigins={mock.props.allBeerOrigins}
      />
    );
  }

  function mocks() {
    mock = {
      props: {
        numItems: 10,
        currentPage: 1,
        onFilterUpdate: jest.fn(),
        onPageChange: jest.fn(),
        allBeerTypes: [],
        allBeerOrigins: []
      }
    };
  }
});
