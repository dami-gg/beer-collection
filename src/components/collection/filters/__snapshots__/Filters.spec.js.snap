// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters should match the snapshot 1`] = `
<div
  className="filters"
>
  <div
    className="search-box filters__filter filters__filter--search"
  >
    <input
      className="search-box__input"
      onChange={[Function]}
      placeholder="Search by name"
      type="text"
      value=""
    />
    <div
      className="search-box__clear"
    />
    <i
      className="search-box__icon fa fa-search"
    />
  </div>
  <div
    className="filters__dropdowns"
  >
    <div
      className="dropdown-filter dropdown-filter--type filters__filter filters__filter--type"
    >
      <button
        className="button button--blue dropdown-filter__button"
        onClick={[Function]}
        type="button"
      >
        All types
        <span
          className="dropdown-filter__arrow"
        />
      </button>
      <ul
        className="options hidden"
      />
    </div>
    <div
      className="dropdown-filter dropdown-filter--origin filters__filter filters__filter--origin"
    >
      <button
        className="button button--blue dropdown-filter__button"
        onClick={[Function]}
        type="button"
      >
        All origins
        <span
          className="dropdown-filter__arrow"
        />
      </button>
      <ul
        className="options hidden"
      />
    </div>
  </div>
  <div
    className="pagination  filters__filter filters__filter--pagination"
  >
    <a
      className="pagination__button
                       
                       "
      onClick={[Function]}
    >
      <span>
        «
      </span>
    </a>
    <a
      className="pagination__button
                       
                       "
      onClick={[Function]}
    >
      <span>
        1
      </span>
    </a>
    <a
      className="pagination__button
                       
                       pagination__button--disabled"
      onClick={[Function]}
    >
      <span>
        ...
      </span>
    </a>
    <a
      className="pagination__button
                       
                       "
      onClick={[Function]}
    >
      <span>
        6
      </span>
    </a>
    <a
      className="pagination__button
                       
                       "
      onClick={[Function]}
    >
      <span>
        7
      </span>
    </a>
    <a
      className="pagination__button
                       
                       "
      onClick={[Function]}
    >
      <span>
        8
      </span>
    </a>
    <a
      className="pagination__button
                       
                       pagination__button--disabled"
      onClick={[Function]}
    >
      <span>
        ...
      </span>
    </a>
    <a
      className="pagination__button
                       
                       "
      onClick={[Function]}
    >
      <span>
        NaN
      </span>
    </a>
    <a
      className="pagination__button
                       
                       "
      onClick={[Function]}
    >
      <span>
        »
      </span>
    </a>
  </div>
</div>
`;
